### PR TITLE
EDU-853 Allow tags and search tokens with less than 3 characters

### PR DESCRIPTION
**Pull request by ahelmberger**
Closes https://educandu.atlassian.net/browse/EDU-853

### CODE REVIEW

File: src/components/search-bar.js [hunk 1]
```diff
@@ -18,10 +16,10 @@ export default function SearchBar({ initialValue, autoFocus, onSearch }) {
   };
 
   const handleSearch = value => {
-    if (value.length < MIN_CHAR_COUNT) {
+    if (!value.length) {
       Modal.error({
         title: t('common:error'),
-        content: t('common:searchTextTooShort', { minCharCount: MIN_CHAR_COUNT }),
+        content: t('common:searchTextTooShort', { minCharCount: 1 }),
```
Comment (adrianaluchian) [inline src/components/search-bar.js:22]: Maybe we should change the validation to say that a value is required. Seems a bit weird and inconsistent this way now.

File: src/services/document-service.js [hunk 2]
```diff
@@ -81,35 +86,52 @@ class DocumentService {
...
+
+    for (const token of tokens) {
+      const isNegative = token.length > 1 && token.startsWith('-');
+      const rawToken = token.substr(isNegative ? 1 : 0).toLowerCase();
+      if (isNegative) {
+        allNegativeTokens.add(rawToken);
+      } else if (rawToken.length < DOCUMENT_PARTIAL_SEARCH_THRESHOLD) {
+        allPositiveTokens.add(rawToken);
+        positiveExactTokens.add(rawToken);
+      } else {
```
Comment (adrianaluchian) [inline src/services/document-service.js:102]: Please bear with my little nitpicking here.
This 3rd case is a little hard to follow in one's mind (or this one's mind at least), you have to do a recap "so not negative, then length >= ..." 😅 
I normally like to avoid cascading `else` blocks as much I try to avoid nested `if` blocks, when possible, but here I think it would make it easier to follow as :
```
else {
     allPositiveTokens.add(rawToken);

     if (rawToken.length < DOCUMENT_PARTIAL_SEARCH_THRESHOLD) {
        positiveExactTokens.add(rawToken);
     } else {
        positivePartialTokens.add(rawToken);
    }
}
````
Oooor... we just set `allPositiveTokens` by joining `positiveExactTokens` and `positivePartialTokens` at the end of the iteration.

File: src/services/document-service.js [hunk 3]
```diff
@@ -81,35 +86,52 @@ class DocumentService {
...
+        allPositiveTokens.add(rawToken);
+        positivePartialTokens.add(rawToken);
+      }
+    }
 
-    if (!positiveTokens.size) {
+    if (!positiveExactTokens.size && !positivePartialTokens.size) {
```
Comment (adrianaluchian) [inline src/services/document-service.js:108]: Basically `!allPositiveTokens.size`, right?

File: src/services/document-service.spec.js [hunk 4]
```diff
@@ -972,7 +972,7 @@ describe('document-service', () => {
         description: 'Description 1',
         slug: 'doc-1',
         sections: [],
-        tags: ['music', 'instructor', 'Dj.D', 'Cretu'],
+        tags: ['music', 'instructor', 'Dj.D', 'Cretu', '1'],
```
Comment (adrianaluchian) [inline src/services/document-service.spec.js:975]: I'm sure of the author of this "Cretu" name in code 😊

File: src/services/document-service.spec.js [hunk 5]
```diff
@@ -1071,7 +1071,7 @@ describe('document-service', () => {
       });
     });
 
-    describe('when I search with a query that returns multiple documents', () => {
+    describe('when the query matches multiple documents', () => {
       it('does not contain archived documents', async () => {
         const results = await sut.getSearchableDocumentsMetadataByTags('music');
         expect(results.map(result => result.title)).not.toContain('Doc 4');
```
Comment (adrianaluchian) [inline src/services/document-service.spec.js:1077]: We've got the test for searching "1 or 2" and the one searching "Doc 4", but I think we need one testing something like "Music 2", to cover the case of both tags matching in one single document, one matching exactly and one partially. This is a case not covered in the current tests as far as I could spot.

File: src/components/search-bar.js (hunk 1 above)
Comment (ahelmberger) [reply src/components/search-bar.js]: Agreed, I created a more specific validation message.

File: src/services/document-service.js (hunk 2 above)
Comment (ahelmberger) [reply src/services/document-service.js:99]: Good idea. I followed your first suggestion and divided the logic into nested binary decisions.

File: src/services/document-service.js (hunk 3 above)
Comment (ahelmberger) [reply src/services/document-service.js]: Yes, you're right (I introduced the `allPositiveTokens` at a later point not realizing that it can also simplify this condition), changed!

File: src/services/document-service.spec.js (hunk 4 above)
Comment (ahelmberger) [reply src/services/document-service.spec.js:975]: It was **not** Maria Magdalena :rofl:

File: src/services/document-service.spec.js (hunk 5 above)
Comment (ahelmberger) [reply src/services/document-service.spec.js:1084]: True. I added a test case with two terms, one of them matching one doc partially, the other one matching a different doc exactly.

File: src/services/document-service.spec.js (hunk 4 above)
Comment (adrianaluchian) [reply src/services/document-service.spec.js:975]: 😂
